traffic_ops/traffic_ops_golang/test/error_context.go

```diff
@@ -0,0 +1,214 @@
+/*
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package test
+
+import "fmt"
+
+type errorCode struct {
+	error
+	cause error
+	code  int
+}
+
+type ErrorCoder interface {
+	Error() string
+	ErrorCode() int
+	Cause() error
+	Prepend(string, ...interface{}) ErrorCoder
+}
+
+func NewError(code int, fmtStr string, fmtArgs ...interface{}) ErrorCoder {
+	err := fmt.Errorf(fmtStr, fmtArgs...)
+	return &errorCode{err, err, code}
+}
+
+func (e errorCode) ErrorCode() int {
+	return e.code
+}
+
+// not a pointer receiver, does not modify itself
+func (e errorCode) Prepend(fmtStr string, fmtArgs ...interface{}) ErrorCoder {
+	err := fmt.Errorf(fmtStr, fmtArgs...)
+	e.error = fmt.Errorf("%v %v", err, e.error)
+	return e
+}
+
+func (e errorCode) Cause() error {
+	return e.cause
+}
+
+func AddErrorCode(code int, err error) ErrorCoder {
+	return NewError(code, "%v", err)
+}
+
+// ErrorContext
+// contains a list of all error codes (a whitelist)
+//		- allows user to make sure they are creating the correct errors
+//		- actually a map
+//			lookup can be done without linear search
+//			we can use the map to keep count of which errors are made
+// contains mapping from error code to name (either for testing metainfo or used in case no args are given)
+//		not required for all error codes, or for any
+//
+// context.NewError was made to improve upon test.NewError
+type ErrorContext struct {
+	calledNewError  bool
+	createdNewError bool
+	doPanic         bool
+	name            string
+	codes           map[uint]uint
```

Comment:
had to trace thru the code to figure out where the whitelist came from, then saw the comment above.  Should this be called "whitelist" to make that apparent?

Reply:
`codes` has a double use. It is both a whitelist of codes and a count of which error codes have been called. There is a `whitelisted` method, which I think communicates well enough when the map of error codes is used as a whitelist in member functions. In all other references besides object construction, the variable is used for its other purpose.

Maybe it would be more clear if the test refers to its code list as a whitelist? Or if I have a comment somewhere else? Because the `whitelisted` method is unexported and the primary purpose of the ErrorContext is to be a whitelist, I would think most people would refer to the ErrorContext comment.

Reply:
>codes has a double use. It is both a whitelist of codes and a count of which error codes have been called.
It would be helpful to add that as a GoDoc comment to the variable, i.e.
```
	// codes is both a whitelist of codes and a count of which error codes have been called.
	codes           map[uint]uint
```